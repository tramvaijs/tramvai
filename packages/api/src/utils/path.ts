import path from 'node:path';

export const resolveAbsolutePathForFile = ({
  file,
  sourceDir,
  rootDir,
}: {
  file: string;
  sourceDir: string;
  rootDir: string;
}): string => {
  if (path.isAbsolute(file)) {
    return file;
  }
  if (path.isAbsolute(sourceDir)) {
    return path.join(sourceDir, file);
  }
  if (path.isAbsolute(rootDir)) {
    return path.join(rootDir, sourceDir, file);
  }
  return path.resolve(rootDir, sourceDir, file);
};

export const resolveAbsolutePathForFolder = ({
  folder,
  rootDir,
}: {
  folder: string;
  rootDir: string;
}): string => {
  if (path.isAbsolute(folder)) {
    return folder;
  }
  if (path.isAbsolute(rootDir)) {
    return path.join(rootDir, folder);
  }
  return path.resolve(rootDir, folder);
};
