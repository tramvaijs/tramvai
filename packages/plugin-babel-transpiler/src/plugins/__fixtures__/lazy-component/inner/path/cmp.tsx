import React from 'react';

const InnerCmp = () => {
  return <div>inner cmp</div>;
}

export default InnerCmp;
