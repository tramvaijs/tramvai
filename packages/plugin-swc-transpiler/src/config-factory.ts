import path from 'node:path';
import { existsSync } from 'node:fs';

import { WebpackTranspilerInputParameters } from '@tramvai/plugin-webpack-builder';
import isEmpty from '@tinkoff/utils/is/empty';
import { sync as resolve } from 'resolve';
import findCacheDir from 'find-cache-dir';
import type { Options as SwcOptions } from '@swc/core';

const TRAMVAI_SWC_TARGET_PATH = '@tramvai/swc-integration/target/wasm32-wasi';

const NOT_SUPPORTED_FIELDS: (keyof WebpackTranspilerInputParameters)[] = [
  'enableFillActionNamePlugin',
];
let warningWasShown = false;

export const configFactory = (config: Partial<WebpackTranspilerInputParameters>): SwcOptions => {
  const {
    env = 'development',
    isServer = false,
    modules = false,
    excludesPresetEnv,
    browsersListTargets,
    typescript = false,
    hot = false,
    removeTypeofWindow,
    tramvai = false,
    rootDir = process.cwd(),
    generateDataQaTag = false,
    // disabled because `arrayLikeIsIterable` and `iterableIsArray` assumtions is not supported yet in swc,
    // this can leads to incorrect code generation e.g. when Set with spread is used - `[...new Set()] => [].concat(new Set())`
    // TCORE-4904
    loose = false,
    externalHelpers = true,
  } = config;

  if (!warningWasShown) {
    for (const field of NOT_SUPPORTED_FIELDS) {
      if (config[field] && !isEmpty(config[field])) {
        console.warn(
          `@tramvai/swc-integration do not support "${field}" configuration. Consider removing it from tramvai.json`
        );

        warningWasShown = true;
      }
    }

    const swcrcPath = path.resolve(rootDir, '.swcrc');

    if (existsSync(swcrcPath)) {
      console.warn(
        `Found .swcrc config in the app root directory ("${swcrcPath}").
Having swc config may conflict with @tramvai/cli configuration`
      );

      warningWasShown = true;
    }
  }

  const resolveWasmFile = (pluginName: string, type: 'debug' | 'release') => {
    return resolve(`${TRAMVAI_SWC_TARGET_PATH}/${type}/${pluginName}.wasm`);
  };

  const resolveTramvaiSwcPlugin = (pluginName: string) => {
    try {
      return resolveWasmFile(pluginName, 'debug');
    } catch (_) {
      try {
        return resolveWasmFile(pluginName, 'release');
      } catch (__) {
        throw new Error(
          `Cannot find tramvai swc-plugin "${pluginName}" related to the "${rootDir}" directory`
        );
      }
    }
  };
  function hasJsxRuntime() {
    try {
      resolve('react/jsx-runtime', { basedir: rootDir });
      return true;
    } catch (e) {
      return false;
    }
  }

  return {
    env: {
      targets: browsersListTargets,
      // Use relevant core-js version, because it affects which polyfills are included
      // https://github.com/swc-project/swc/blob/main/crates/swc_ecma_preset_env/data/core-js-compat/modules-by-versions.json
      coreJs: require('core-js/package.json').version,
      // disabled because `arrayLikeIsIterable` and `iterableIsArray` assumtions is not supported yet in swc,
      // this can leads to incorrect code generation e.g. when Set with spread is used - `[...new Set()] => [].concat(new Set())`
      // TCORE-4904
      loose,
      exclude: excludesPresetEnv,
      mode: 'entry',
    },
    module: {
      type: modules || 'es6',
    },
    isModule: 'unknown',
    jsc: {
      externalHelpers,
      parser: {
        syntax: typescript ? 'typescript' : 'ecmascript',
        decorators: true,
        tsx: true,
        jsx: true,
        exportDefaultFrom: true,
      },
      transform: {
        legacyDecorator: true,
        react: {
          runtime: hasJsxRuntime() ? 'automatic' : 'classic',
          development: env === 'development',
          refresh: hot && env === 'development' && !isServer,
        },
        optimizer: {
          globals: {
            // let the webpack replace NODE_ENV as replacement with swc may mess up with tests
            envs: [],
            typeofs: removeTypeofWindow
              ? {
                  window: isServer ? 'undefined' : 'object',
                }
              : {},
          },
        },
      },
      experimental: {
        cacheRoot: findCacheDir({ cwd: rootDir, name: 'swc' }),
        plugins: [
          [resolveTramvaiSwcPlugin('create_token_pure'), {}],
          [resolveTramvaiSwcPlugin('lazy_component'), {}],
          isServer && [resolveTramvaiSwcPlugin('dynamic_import_to_require'), {}],
          tramvai && env === 'development' && [resolveTramvaiSwcPlugin('provider_stack'), {}],
          generateDataQaTag && [resolveTramvaiSwcPlugin('react_element_info_unique'), {}],
        ].filter(Boolean) as Array<[string, Record<string, any>]>,
      },
    },
  };
};
