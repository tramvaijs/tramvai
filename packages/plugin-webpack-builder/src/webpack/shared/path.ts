import path from 'node:path';

export const resolveAbsolutePathForEntry = ({
  entry,
  sourceDir,
  rootDir,
}: {
  entry: string;
  sourceDir: string;
  rootDir: string;
}): string => {
  if (path.isAbsolute(entry)) {
    return entry;
  }
  if (path.isAbsolute(sourceDir)) {
    return path.join(sourceDir, entry);
  }
  if (path.isAbsolute(rootDir)) {
    return path.join(rootDir, sourceDir, entry);
  }
  return path.resolve(rootDir, sourceDir, entry);
};

export const resolveAbsolutePathForOutput = ({
  output,
  rootDir,
}: {
  output: string;
  rootDir: string;
}): string => {
  if (path.isAbsolute(output)) {
    return output;
  }
  if (path.isAbsolute(rootDir)) {
    return path.join(rootDir, output);
  }
  return path.resolve(rootDir, output);
};
