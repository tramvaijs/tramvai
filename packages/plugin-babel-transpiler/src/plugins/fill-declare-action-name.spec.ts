import path from 'path';
import pluginTester from 'babel-plugin-tester';
import { configFactory } from '../config-factory';

pluginTester({
  plugin: {}, // плагин уже есть в основном конфиге, если подключить ещё тут то babel упадёт
  pluginName: 'fill-declare-action-name',
  filename: path.join(__dirname, '__fixtures__', 'fill-declare-action-name', 'test'),
  babelOptions: configFactory({
    typescript: true,
    generateDataQaTag: false,
    loader: false,
    enableFillDeclareActionNamePlugin: true,
  }),
  tests: {
    'import: fill absent name': {
      fixture: 'import-mixed.ts',
      snapshot: true,
    },
    'import: do nothing': {
      fixture: 'import-with-name.ts',
      snapshot: true,
    },
    'import: fill all actions': {
      fixture: 'import-without-name.ts',
      snapshot: true,
    },
    'require: do nothing': {
      fixture: 'require-with-name.ts',
      snapshot: true,
    },
    'require: fill all actions': {
      fixture: 'require-without-name.ts',
      snapshot: true,
    },
  },
});
