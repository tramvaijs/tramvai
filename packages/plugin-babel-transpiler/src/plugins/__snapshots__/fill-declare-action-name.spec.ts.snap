// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fill-declare-action-name import: do nothing: import: do nothing 1`] = `

import { declareAction } from '@tramvai/core';

const action = declareAction({
  name: 'action',
  fn: () => {
    console.log('action');
  },
});

const secondAction = declareAction({
  name: 'secondAction',
  fn: () => {
    console.log('second action');
  },
  conditions: {
    onlyBrowser: true,
  },
});

export default declareAction({
  name: 'anonymousAction',
  fn: () => {
    console.log('anonymous action');
  },
  conditions: {
    onlyBrowser: true,
  },
});

      ↓ ↓ ↓ ↓ ↓ ↓

import { declareAction } from '@tramvai/core';
const action = declareAction({
  name: 'action',
  fn: () => {
    console.log('action');
  },
});
const secondAction = declareAction({
  name: 'secondAction',
  fn: () => {
    console.log('second action');
  },
  conditions: {
    onlyBrowser: true,
  },
});
export default declareAction({
  name: 'anonymousAction',
  fn: () => {
    console.log('anonymous action');
  },
  conditions: {
    onlyBrowser: true,
  },
});


`;

exports[`fill-declare-action-name import: fill absent name: import: fill absent name 1`] = `

import { declareAction } from '@tramvai/core';

const action = declareAction({
  name: 'action',
  fn: () => {
    console.log('action');
  },
});

const secondAction = declareAction({
  fn: () => {
    console.log('second action');
  },
  conditions: {
    onlyBrowser: true,
  },
});

      ↓ ↓ ↓ ↓ ↓ ↓

import { declareAction } from '@tramvai/core';
const action = declareAction({
  name: 'action',
  fn: () => {
    console.log('action');
  },
});
const secondAction = declareAction({
  name: 'secondAction__-686ypu',
  fn: () => {
    console.log('second action');
  },
  conditions: {
    onlyBrowser: true,
  },
});


`;

exports[`fill-declare-action-name import: fill all actions: import: fill all actions 1`] = `

import { declareAction } from '@tramvai/core';

const action = declareAction({
  fn: () => {
    console.log('action');
  },
});

const secondAction = declareAction({
  fn: () => {
    console.log('second action');
  },
  conditions: {
    onlyBrowser: true,
  },
});

export default declareAction({
  fn: () => {
    console.log('anonymous action');
  },
  conditions: {
    onlyBrowser: true,
  },
});

      ↓ ↓ ↓ ↓ ↓ ↓

import { declareAction } from '@tramvai/core';
const action = declareAction({
  name: 'action__-1vwsoc',
  fn: () => {
    console.log('action');
  },
});
const secondAction = declareAction({
  name: 'secondAction__-1vsyqf',
  fn: () => {
    console.log('second action');
  },
  conditions: {
    onlyBrowser: true,
  },
});
export default declareAction({
  name: 'import-without-name.ts__ciqoha',
  fn: () => {
    console.log('anonymous action');
  },
  conditions: {
    onlyBrowser: true,
  },
});


`;

exports[`fill-declare-action-name require: do nothing: require: do nothing 1`] = `

// @ts-ignore
const { declareAction } = require('@tramvai/core');

// @ts-ignore
const action = declareAction({
  name: 'action',
  fn: () => {
    console.log('action');
  },
});

// @ts-ignore
const secondAction = declareAction({
  name: 'secondAction',
  fn: () => {
    console.log('second action');
  },
  conditions: {
    onlyBrowser: true,
  },
});

      ↓ ↓ ↓ ↓ ↓ ↓

// @ts-ignore
const { declareAction } = require('@tramvai/core');

// @ts-ignore
const action = declareAction({
  name: 'action',
  fn: () => {
    console.log('action');
  },
});

// @ts-ignore
const secondAction = declareAction({
  name: 'secondAction',
  fn: () => {
    console.log('second action');
  },
  conditions: {
    onlyBrowser: true,
  },
});


`;

exports[`fill-declare-action-name require: fill all actions: require: fill all actions 1`] = `

// @ts-ignore
const { declareAction } = require('@tramvai/core');

// @ts-ignore
const action = declareAction({
  fn: () => {
    console.log('action');
  },
});

// @ts-ignore
const secondAction = declareAction({
  fn: () => {
    console.log('second action');
  },
  conditions: {
    onlyBrowser: true,
  },
});

      ↓ ↓ ↓ ↓ ↓ ↓

// @ts-ignore
const { declareAction } = require('@tramvai/core');

// @ts-ignore
const action = declareAction({
  name: 'action__qfz4ju',
  fn: () => {
    console.log('action');
  },
});

// @ts-ignore
const secondAction = declareAction({
  name: 'secondAction__-wofu95',
  fn: () => {
    console.log('second action');
  },
  conditions: {
    onlyBrowser: true,
  },
});


`;
